### DISCUSSION TITLE
Thread safety

### DISCUSSION THREAD

**Discussion by nashysolutions** [pointfreeco/swift-dependencies#63] [Q&A]
Is this the recommended way (or one of the recommended ways) for making dependencies thread safe?

```swift
private let lock = NSLock()

public struct CacheDependencyKey: DependencyKey {
    
    private static var _cache = Cache<RemoteImageResource>(maxSize: 100)
    
    public static var liveValue: Cache<RemoteImageResource> {
        get { lock.withCriticalSection { _cache } }
        set { lock.withCriticalSection { _cache = newValue } }
    }
}

public extension DependencyValues {
    
    var resourceCache: Cache<RemoteImageResource> {
        get { self[CacheDependencyKey.self] }
        set { self[CacheDependencyKey.self] = newValue }
    }
}

private extension NSLocking {
    
    func withCriticalSection<T>(block: () throws -> T) rethrows -> T {
        lock()
        defer { unlock() }
        return try block()
    }
}
```

**Comment by mbrandonw** [comment]
Locks definitely work and there's no reason to not use them. You can also hold onto an actor under the hood of the dependency and use that for synchronization. Note that you generally do not want to make the dependency itself an actor, but you can still use one for the live implementation.

However, the are some problems with your dependency as you have demonstrated it above. You shouldn't think of the cache _itself_ as the dependency, but rather _access_ to the cache as the dependency. That way you can lock the beginning-to-end process of accessing the cache and grabbing a value. Right now you are only locking the access to the cache, but not fetching/setting data.

One way to do this is to design a client with `get`/`set` endpoints:

```swift
struct CacheClient {
  var get: (String) -> Image
  var set: (String, Image) -> Void
}
extension CacheClient: DependencyKey {
  static var liveValue: CacheClient {
    let lock = NSLock()
    let cache = Cache<RemoteImageResource>(maxSize: 100)
    return Self(
      get: { key in lock.withCriticalSection { cache.get(key) } },
      set: { key, image in lock.withCriticalSection { cache.set(key, image) } }
    )
  }
}
```

**Reply by nashysolutions** [reply]
Thank you 🙏
